language/extensions/move-table-extension/sources/Table.move

```diff
@@ -0,0 +1,93 @@
+/// Type of large-scale storage tables.
+module Extensions::Table {
+    /// Type of tables
+    struct Table<phantom K, phantom V> has store {
+        handle: u128
+    }
+
+    /// Create a new Table.
+    public fun new<K, V: store>(): Table<K, V> {
+        Table{handle: new_table_handle()}
+    }
+
+    /// Destroy a table. The table must be empty to succeed.
+    public fun destroy_empty<K, V>(table: Table<K, V>) {
+        destroy_empty_box<K, V, Box<V>>(&table);
+        drop_unchecked_box<K, V, Box<V>>(table)
+    }
+
+    /// Insert the pair (`key`, `val`) to `table`.
+    /// Aborts if there is already an entry for `key`.
+    public fun insert<K, V>(table: &mut Table<K, V>, key: &K, val: V) {
+        insert_box<K, V, Box<V>>(table, key, Box{val})
+    }
+
+    /// Acquire an immutable reference to the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    public fun borrow<K, V>(table: &Table<K, V>, key: &K): &V {
+        &borrow_box<K, V, Box<V>>(table, key).val
+    }
+
+    /// Acquire a mutable reference to the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    public fun borrow_mut<K, V>(table: &mut Table<K, V>, key: &K): &mut V {
+        &mut borrow_box_mut<K, V, Box<V>>(table, key).val
+    }
+
+    /// Returns the length of the table, i.e. the number of entries.
+    public fun length<K, V>(table: &Table<K, V>): u64 {
+        length_box<K, V, Box<V>>(table)
+    }
+
+    /// Returns true if this table is empty.
+    public fun empty<K, V>(table: &Table<K, V>): bool {
+        length(table) == 0
+    }
+
+    /// Acquire a mutable reference to the value which `key` maps to.
+    /// Insert the pair (`key`, `default`) first if there is no entry for `key`.
+    public fun borrow_mut_with_default<K: copy + drop, V: drop>(table: &mut Table<K, V>, key: &K, default: V): &mut V {
+        if (!contains(table, key)) {
+            insert(table, copy key, default)
+        };
+        borrow_mut(table, key)
+    }
+
+    /// Remove from `table` and return the value which `key` maps to.
+    /// Aborts if there is no entry for `key`.
+    public fun remove<K, V>(table: &mut Table<K, V>, key: &K): V {
+        let Box{val} = remove_box<K, V, Box<V>>(table, key);
+        val
+    }
+
+    /// Returns true iff `table` contains an entry for `key`.
+    public fun contains<K, V>(table: &Table<K, V>, key: &K): bool {
+        contains_box<K, V, Box<V>>(table, key)
+    }
+
+    #[test_only]
+    /// Testing only: allows to drop a table even if it is not empty.
+    public fun drop_unchecked<K, V>(table: Table<K, V>) {
```

Comment:
I think it will be useful for testing so you can run the code without Move complaining about not consuming the table handle. 

If I follow correctly, this native function is a no-op and `destroy_empty_box` is where the real logic is

Reply:
It's used in the current testsuite, as Todd describes.